src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/step/forcemerge/WaitForForceMergeStep.kt

```diff
@@ -0,0 +1,145 @@
+) : Step(name, managedIndexMetaData) {
+
+    private val logger = LogManager.getLogger(javaClass)
+    private var stepStatus = StepStatus.STARTING
+    private var info: Map<String, Any>? = null
+
+    @Suppress("TooGenericExceptionCaught") // TODO see if we can refactor to catch GenericException in Runner.
+    override suspend fun execute() {
+        val indexName = managedIndexMetaData.index
+
+        logger.info("Waiting for force merge to complete on [$indexName]")
+
+        // Retrieve maxNumSegments value from ActionProperties. If ActionProperties is null, update failed info and return early.
+        val maxNumSegments = getMaxNumSegments() ?: return
+
+        // Get the number of shards with a segment count greater than maxNumSegments, meaning they are still merging
+        val shardsStillMergingSegments = getShardsStillMergingSegments(indexName, maxNumSegments)
+        // If shardsStillMergingSegments is null, failed info has already been updated and can return early
+        shardsStillMergingSegments ?: return
+
+        // If there are no longer shardsStillMergingSegments, then the force merge has completed
+        if (shardsStillMergingSegments == 0) {
+            logger.info("Force merge completed on [$indexName]")
+
+            stepStatus = StepStatus.COMPLETED
+            info = mapOf("message" to "Force merge completed")
+        } else {
+            /*
+             * If there are still shards with segments merging then no action is taken and the step will be reevaluated
+             * on the next run of the ManagedIndexRunner.
+             *
+             * However, if an internal timeout of 12 hours is reached, the force_merge action will be marked as failed.
+             * This is for cases where the force merge stops silently, such as when nodes relocate during a force
+             * merge operation. Without a timeout, the segment count would stop going down as merging would no longer
+             * occur and the managed index would become stuck in this action.
+             */
+            val timeWaitingForForceMerge: Duration = Duration.between(getStepStartTime(), Instant.now())
+            if (timeWaitingForForceMerge.toSeconds() > FORCE_MERGE_TIMEOUT_IN_SECONDS) {
+                logger.error(
+                    "Force merge on [$indexName] timed out with [$shardsStillMergingSegments] shards containing unmerged segments"
+                )
+
+                stepStatus = StepStatus.FAILED
+                info = mapOf("message" to "Force merge timed out")
+            } else {
+                logger.debug(
+                    "Force merge still running on [$indexName] with [$shardsStillMergingSegments] shards containing unmerged segments"
+                )
+
+                stepStatus = StepStatus.CONDITION_NOT_MET
+                info = mapOf("message" to "Waiting for force merge to complete")
+            }
+        }
+    }
+
+    private fun getMaxNumSegments(): Int? {
```

Comment:
np: this is hard to read, something like this might be easier:
```
private fun getMaxNumSegments(): Int? {
    val actionProperties = managedIndexMetaData.actionMetaData?.actionProperties
    if (actionProperties == null) {
        stepStatus = StepStatus.FAILED
        info = mapOf("message" to "Unable to retrieve [${ActionProperties.MAX_NUM_SEGMENTS}] from ActionProperties=$actionProperties")
    }
    return actionProperties?.getInt(ActionProperties.MAX_NUM_SEGMENTS)
    }
```

Reply:
Agreed, it was an Intellij suggestion but did make the code harder to read.